### ISSUE TITLE
BC37255: Combined length of user strings used by the program exceeds allowed limit. Try to decrease use of string or XML literals

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/aspnetcore#21787]
I receive this error on my datasets since I've installed the latest VS2017. I have big tables, and think the MSDataSetGenerator generates code different, so now the compiler can't handle the big generated strings anymore?

There is not much information about the error message BC37255.

Can you help?

_This issue has been moved from https://developercommunity.visualstudio.com/content/problem/192728/bc37255-combined-length-of-user-strings-used-by-th.html
VSTS ticketId: 561492_
_These are the original issue comments:_
(no comments)
_These are the original issue solutions:_
(no solutions)

**Comment by pranavkm** [comment]
> might have produced an invalid assembly if the total string size exceeded the limits allows by ECMA335 specification.

@tmat \ @gafter out of curiosity, what is this size limit?

**Comment by tmat** [comment]
2^24 bytes. Literals are UTF16 encoded, so 8M characters.

**Comment by gr8tushar** [comment]
Hi All,
We are stuck in the middle. For one of our solution, we have migrated most of the apps to .NET Core. But this issue is preventing us to go live with all the apps in .NET Core. 

It has become a night mare and comes with huge cost in maintaining two versions of the system.

Awaiting some positive solution.

**Comment by pranavkm** [comment]
The right fix for this would involve razor codegeneration to stash string literals in to a resx file and changing WriteLiteral calls to use these resx strings. Implementing this is going to be a fair bit of work - the code generation would need to add this feature and we would need to change the SDK to express multiple outputs produced per cshtml file.

In the meanwhile, app developers have a few options to work around this:

a) Consider manually stashing literals in to resx files. This would follow the same pattern as using [view localization](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/localization?view=aspnetcore-3.1#view-localization) and it's likely you could repurpose the existing localization APIs for this.

b) Another alternative would be to extracting large chunks of HTML in to separate files and emitting these out at runtime:
  - https://github.com/pranavkm/IncludeTagHelper/blob/master/app/Views/Home/Index.cshtml#L10
  - https://github.com/pranavkm/IncludeTagHelper/blob/master/app/Views/Home/Index.Content.html
  - https://github.com/pranavkm/IncludeTagHelper/blob/master/taghelper/IncludeHtmlTagHelper.cs

c) Yet another option that's already been described here is to turn off build time compilation and rely on runtime compilation. Runtime compilation compiles individual views in to separate assembly. This makes it less likely to reach the 8M limit for strings. To disable build time compilation, 

```xml
<PropertyGroup>
   <RazorCompileOnBuild>false</RazorCompileOnBuild>
</PropertyGroup>
```

Follow https://docs.microsoft.com/en-us/aspnet/core/mvc/views/view-compilation?view=aspnetcore-3.1&tabs=visual-studio#enable-runtime-compilation-in-an-existing-project to learn about enabling runtime compilation.

**Comment by SayedMahmoudGomaa** [comment]
I got this error too on a desktop application project, I'm using VS 2019 with .Net framework 4.5.2, can anyone help?

**Comment by luomingui** [comment]
`<PropertyGroup>
    <TargetFramework>netcoreapp2.0</TargetFramework>
    <PublishWithAspNetCoreTargetManifest>false</PublishWithAspNetCoreTargetManifest>
    <MvcRazorCompileOnPublish>false</MvcRazorCompileOnPublish>
  </PropertyGroup>`

**Comment by meyerjoe** [comment]
Got the same issue while migrating a large .net framework MVC app to .net core 6

This issue is blocking further migration.

The workaround with the settings in .csproj doesn't work.

IDE: JetBrains Rider on MacOS

**Comment by ashraqat-sheta** [comment]
i got this error in vs 2022 and .net core6.
the project was working well with vs 2019 .net core 3 but after i upgraded my project to .net6 it broke .
i faced this error many times before but the problem now is bigger.  i really need to know what causes this bug and how to fix it . any help?

**Comment by fxanimal** [comment]
This same error also happened in my project.
Someone from MS told me it's because there are too many views or razor pages. Therefore I split my project into several smaller projects and it worked well.
But the problem is how to call another project's controller action from the main project.
And all projects have to run simultaneously.

**Comment by jumpingjackson** [comment]
Also have the same issue migrating a large .net framework MVC app to .net core 6.

**Comment by DamianEdwards** [comment]
@tmat does the same limit apply to the Blob Heap? I'm imagining what if we were to encode the string literals at compile time to UTF8 byte literals? That might double the amount encoded at least.

**Comment by tmat** [comment]
Yes, it does. I'd recommend generated them as managed resources, if possible.

**Comment by jumpingjackson** [comment]
The conversion to single file .dll as part of the .Net 6 HotReload makes this problem worse correct?  As in someone who might have been fine on .Net 5 with a separate .Views.dll file could be in for a surprise if they upgrade.

**Comment by DamianEdwards** [comment]
@jumpingjackson possibly, although the problem is far more pronounced when moving from non-Core ASP.NET where the views were typically runtime-compiled into separate assemblies.

**Comment by DamianEdwards** [comment]
I [published a repo](https://github.com/DamianEdwards/LargeRazorStringLiterals) with a project that reproduces the issue synthetically (i.e. it generates lots of .cshtml files with lots of literal content).

It also includes an idea for a potential workaround using C# 11 Utf8 string literals but the real-world usability of that is likely pretty poor (all literal content must be passed to a function as a Utf8 string literal).
